lib/backend/syncersv1/updateprocessors/configurationprocessor_test.go

```diff
@@ -0,0 +1,625 @@
+		checkExpectedConfigs(
+			kvps,
+			isGlobalBgpConfig,
+			numBgpConfigs,
+			expected,
+		)
+
+		By("validating a partial configuration")
+		n2n = false
+		res.Spec.LogSeverityScreen = "debug"
+		res.Spec.ASNumber = nil
+		res.Spec.NodeToNodeMeshEnabled = &n2n
+		expected = map[string]interface{}{
+			"loglevel":  "debug",
+			"as_num":    nil,
+			"node_mesh": "{\"enabled\":false}",
+		}
+		kvps, err = cc.Process(&model.KVPair{
+			Key:   globalBgpConfigKey,
+			Value: res,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		checkExpectedConfigs(
+			kvps,
+			isGlobalBgpConfig,
+			numBgpConfigs,
+			expected,
+		)
+
+		By("validating node specific configuration")
+		n2n = false
+		res.Spec.LogSeverityScreen = "debug"
+		res.Spec.ASNumber = nil
+		res.Spec.NodeToNodeMeshEnabled = nil
+		expected = map[string]interface{}{
+			"loglevel":  "debug",
+			"as_num":    nil,
+			"node_mesh": nil,
+		}
+		kvps, err = cc.Process(&model.KVPair{
+			Key:   nodeBgpConfigKey,
+			Value: res,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		checkExpectedConfigs(
+			kvps,
+			isNodeBgpConfig,
+			numBgpConfigs,
+			expected,
+		)
+	})
+
+	It("should handle node cluster information", func() {
+		cc := updateprocessors.NewClusterInfoUpdateProcessor()
+		res := apiv2.NewClusterInformation()
+
+		By("validating an empty per node cluster is processed correctly")
+		kvps, err := cc.Process(&model.KVPair{
+			Key:   nodeClusterKey,
+			Value: res,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		checkExpectedConfigs(
+			kvps,
+			isNodeFelixConfig,
+			numClusterConfigs,
+			nil,
+		)
+
+		By("validating it is not possible to set/override values using the annotations")
+		res.Annotations = map[string]string{
+			"config.projectcalico.org/ClusterType": "this is not validated!",
+			"config.projectcalico.org/NewField":    "this is also not validated!",
+		}
+		kvps, err = cc.Process(&model.KVPair{
+			Key:   nodeClusterKey,
+			Value: res,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		checkExpectedConfigs(
+			kvps,
+			isNodeFelixConfig,
+			numClusterConfigs,
+			nil,
+		)
+	})
+})
+
+// Check the KVPairs returned by the UpdateProcessor are as expected.  The expectedValues contains
+// the expected set of data in the updates, any config not specified in the set is expected
+// to be nil in the KVPair.
+// You can use expectedValues to verify certain fields were included in the response even
+// if the values were nil.
+func checkExpectedConfigs(kvps []*model.KVPair, dataType int, expectedNum int, expectedValues map[string]interface{}) {
+	// Copy/convert input data.  We keep track of:
+	// - all field names, so that we can check for duplicates
+	// - extra fields that we have not yet seen
+	// - expected field values that we have not yet validated
+	ev := make(map[string]interface{}, len(expectedValues))
+	for k, v := range expectedValues {
+		ev[k] = v
+	}
+	allNames := map[string]struct{}{}
+
+	By(" - checking the expected number of results")
```

Comment:
it made the output a lot more readable.

since these sets of "sub tests" are repeated for a lot of the tests - so it's easier to see what the main test was.